### PR TITLE
Cron scaler: Fix unexpected instance count change

### PR DESCRIPTION
**Pull request by zou2699**
fix IsActive in time boundary

Signed-off-by: tux <zou2699@gmail.com>

_Provide a description of what has been changed_

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [ ] Tests have been added
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [ ] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes #
fix IsActive in time boundary 
for example, when we use this cron configuration to scale pod at the last day of a month
```
start: 0 8 28-31 * *
end: 0 20 1 * *
```
if runTime is 2022-01-30T7:59:59.9999999 
might happen
nextStartTime 2022-01-30T8:00:00.000000
currentTime 2022-01-30T8:00:00.000111
at this time  `currentTime < nextStartTime` will be false，IsActive will return true，trigger unexpected scaler

Fixes: https://github.com/kedacore/keda/issues/3854

### CODE REVIEW

**Comment by JorTurFer** [comment]
Hey,
What is the problem you are trying to solve? I mean, even the description is true, doing that change moves this "issue" to the end time instead of starting time. I think that you shouldn't expect millisecond accuracy in the cron scaler, because it's evaluated every X seconds, indeed.
The scenario you described is correct, but the trigger will be activated/deactivated in the next cycle, usually in 30 seconds

**Comment by zou2699** [comment]
We use the following expression to expand the capacity from the end of 8:00 am on the last day of each month to 8:00 pm on the 1st of the next month
````
start: "0 8 L * *"
end: "0 20 1 * *"
````
But the preexisting dependency library cron does not support L (the last day of month), see issue https://github.com/robfig/cron/issues/464

So we use the following expression instead
````
start:"0 8 28,29,30,31 * *"
end: "0 20 1 * *"
````
But currently cronScaler.IsActive may trigger an issue like
````go
     // if  execTime=2022.10.31 07:59:59.999
      nextStartTime, startTimecronErr := getCronTime(location, s.metadata.start)
      // got nextStartTime=2022.10.31 08:00:00.000
       ......
      nextEndTime, endTimecronErr := getCronTime(location, s.metadata.end)
     // got nextEndTime=2022.11.01 20:00:00.000
      ......
     currentTime := time.Now().Unix()
     // got currentTime=2022.10.31 08:00:00.111
     switch {
       case nextStartTime < nextEndTime && currentTime < nextStartTime:
            return false, nil
       // will use this case, return true, will scale up pod
      // but in the next cycle,  will retrun false       
       case currentTime <= nextEndTime:
           return true, nil
       default:
           return false, nil
}
````
this will cause an unexpected scale up event

when we get the currentTime before getting nextStartTime in cronScaler.IsActive, we can avoid this mistake

**Comment by zroubalik** [comment]
@zou2699 could you please open an issue first, with clear explanation of the problem? Thanks!

**Comment by zroubalik** [comment]
@zou2699 do you see any usecases (that don't suffer from this edge case) where this fix could break the existing behavior?

**Comment by zou2699** [comment]
@zroubalik This will not affect other usecases

**Comment by JorTurFer** [comment]
I think that the change doesn't change the behavior, more than changing the edge case of 59:999. It's true that the opposite case (00.001) will change, but as I said, I think that it's a really "forced" edge case

**Comment by JorTurFer** [comment]
BTW @zou2699 , are you sure that the problem is related with this? I mean, what pooling interval do you have? Because if you are not checking ever 1 second (and even in that case), I'd say that it's really complex to happen

**Comment by zou2699** [comment]
The generation of nextStartTime depends on the current time point, but the currentTime here is obtained after nextStartTime is generated, which will lead to the following problems

For example, in the above example, we don't want an unexpected instance count change at `2022.10.30 07:59:59.999`, we only expect to instance count change at `2022.10.31 08:00:00.000 -> 2022.11.01 20:00:00.000` .

Through this adjustment, the nextStartTime of the time that is not the last day (October 28, October 29, October 30) is always greater than the currentTime:
```
case1:
currentTime: 2022-10-28 07:59:59:999
nextStartTime: 2022-10-28 08:00:000
case2:
currentTime: 2022-10-28 08:00:00:001
nextStartTime: 2022-10-29 08:00:000
```
to avoid unexpected instance count changes, thus supporting 'L' in Cron expressions

**Comment by zroubalik** [comment]
@zou2699 could you please fix the static check problems and add entry to the changelog, we can merge it then.

File: CHANGELOG.md [hunk 1]
```diff
@@ -38,6 +38,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 
 ### New
 
+- **Cron Scaler**: Fix unexpected instance count change. ([#3838](https://github.com/kedacore/keda/pull/3838))
```
Comment (zroubalik) [inline CHANGELOG.md:41]: please move this to the improvements section and sort alphabetically

**Review by zroubalik — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -52,6 +52,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **AWS Scalers**: Add setting AWS endpoint url. ([#3337](https://github.com/kedacore/keda/issues/3337))
 - **Azure Service Bus Scaler**: Add support for Shared Access Signature (SAS) tokens for authentication. ([#2920](https://github.com/kedacore/keda/issues/2920))
 - **Azure Service Bus Scaler:** Support regex usage in queueName / subscriptionName parameters. ([#1624](https://github.com/kedacore/keda/issues/1624))
+- **Cron Scaler**: Fix unexpected instance count change. ([#3838](https://github.com/kedacore/keda/pull/3838))
```
[the remark this answers is not in the document]
Comment (zroubalik) [reply CHANGELOG.md:55]: Move this to ### Improvements section down

**Review by zroubalik — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (zroubalik) [reply CHANGELOG.md:55]: @zou2699 please look at this^

**Pull request merged**